launcher/InstanceImportTask.cpp

```diff
@@ -108,76 +108,79 @@ void InstanceImportTask::downloadFromUrl()
     MMCZip::ArchiveReader packZip(m_archivePath);
-    if (!packZip.collectFiles()) {
-        emitFailed(tr("Unable to open supplied modpack zip file."));
-        return;
-    }
-
     qDebug() << "Attempting to determine instance type";
 
     QString root;
-
+    auto detectInstance = [this, &extractDir, &root](MMCZip::ArchiveReader::File* f, bool& stop) {
+        if (!isRunning()) {
+            stop = true;
+            return true;
+        }
+        auto fileName = f->filename();
+        if (fileName == "/modrinth.index.json") {
+            // process as Modrinth pack
+            qDebug() << "Modrinth:" << true;
+            m_modpackType = ModpackType::Modrinth;
+            stop = true;
+        } else if (fileName == "/bin/modpack.jar" || fileName == "/bin/version.json") {
+            // process as Technic pack
+            qDebug() << "Technic:" << true;
+            extractDir.mkpath("minecraft");
+            extractDir.cd("minecraft");
+            m_modpackType = ModpackType::Technic;
+            stop = true;
+        } else {
+            QFileInfo fileInfo(fileName);
+            if (fileInfo.fileName() == "instance.cfg") {
+                qDebug() << "MultiMC:" << true;
+                m_modpackType = ModpackType::MultiMC;
+                root = cleanPath(fileInfo.path());
+                stop = true;
+                return true;
+            }
+            if (fileInfo.fileName() == "manifest.json") {
+                qDebug() << "Flame:" << true;
+                m_modpackType = ModpackType::Flame;
+                root = cleanPath(fileInfo.path());
+                stop = true;
+                return true;
+            }
+        }
+        QCoreApplication::processEvents();
+        return true;
+    };
     // NOTE: Prioritize modpack platforms that aren't searched for recursively.
     // Especially Flame has a very common filename for its manifest, which may appear inside overrides for example
     // https://docs.modrinth.com/docs/modpacks/format_definition/#storage
-    if (packZip.exists("/modrinth.index.json")) {
-        // process as Modrinth pack
-        qDebug() << "Modrinth:" << true;
-        m_modpackType = ModpackType::Modrinth;
-    } else if (packZip.exists("/bin/modpack.jar") || packZip.exists("/bin/version.json")) {
-        // process as Technic pack
-        qDebug() << "Technic:" << true;
-        extractDir.mkpath("minecraft");
-        extractDir.cd("minecraft");
-        m_modpackType = ModpackType::Technic;
-    } else {
-        root = getRootFromZip(packZip.getFiles());
-        setDetails("");
+    if (!packZip.parse(detectInstance)) {
+        emitFailed(tr("Unable to open supplied modpack zip file."));
+        return;
     }
     if (m_modpackType == ModpackType::Unknown) {
         emitFailed(tr("Archive does not contain a recognized modpack type."));
         return;
     }
     setStatus(tr("Extracting modpack"));
+    setDetails("");
```

Comment:
Is this actually required? I don't think I've seen it before in the codebase...and it's kinda pointless